zson/analyzer.go

```diff
@@ -0,0 +1,699 @@
+	var alias *zng.TypeAlias
+	if zng.IsIdentifier(name) {
+		var err error
+		if alias, err = zctx.LookupTypeAlias(name, typ); err != nil {
+			return nil, err
+		}
+		if existing, ok := a[name]; ok && existing != alias {
+			return nil, fmt.Errorf("type '%s' redefined from '%s' to '%s'", name, existing.ZSON(), typ.ZSON())
+		}
+		typ = alias
+	}
+	a[name] = typ
+	return alias, nil
+}
+
+func (a Analyzer) convertAny(zctx *resolver.Context, val ast.Any, cast zng.Type) (Value, error) {
+	// If we're casting something to a union, then the thing inside needs to
+	// describe itself and we can convert the inner value to a union value when
+	// we know its type (so we can code the selector).
+	if union, ok := cast.(*zng.TypeUnion); ok {
+		v, err := a.convertAny(zctx, val, nil)
+		if err != nil {
+			return nil, err
+		}
+		return a.convertUnion(zctx, v, union)
+	}
+	switch val := val.(type) {
+	case *ast.Primitive:
+		return a.convertPrimitive(zctx, val, cast)
+	case *ast.Record:
+		return a.convertRecord(zctx, val, cast)
+	case *ast.Array:
+		return a.convertArray(zctx, val, cast)
+	case *ast.Set:
+		return a.convertSet(zctx, val, cast)
+	case *ast.Enum:
+		return a.convertEnum(zctx, val, cast)
+	case *ast.Map:
+		return a.convertMap(zctx, val, cast)
+	case *ast.TypeValue:
+		return a.convertTypeValue(zctx, val, cast)
+	}
+	return nil, fmt.Errorf("internal error: unknown ast type in Analyzer.convertAny(): %T", val)
+}
+
+func (a Analyzer) convertPrimitive(zctx *resolver.Context, val *ast.Primitive, cast zng.Type) (Value, error) {
+	typ := zng.LookupPrimitive(val.Type)
+	if typ == nil {
+		return nil, fmt.Errorf("no such primitive type: %s", val.Type)
+	}
+	isNull := typ == zng.TypeNull
+	if cast != nil {
+		// The parser emits Enum values for identifiers but not for
+		// string enum names.  Check if the cast type is an enum,
+		// and if so, convert the string to its enum counterpart.
+		if v := stringToEnum(val, cast); v != nil {
+			return v, nil
+		}
+		var err error
+		typ, err = castType(typ, cast)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if isNull {
+		return &Null{Type: typ}, nil
+	}
+	return &Primitive{Type: typ, Text: val.Text}, nil
+}
+
+func stringToEnum(val *ast.Primitive, cast zng.Type) Value {
+	if enum, ok := cast.(*zng.TypeEnum); ok {
+		if val.Type == "string" {
+			return &Enum{
+				Type: enum,
+				Name: val.Text,
+			}
+		}
+	}
+	return nil
+}
+
+func castType(typ, cast zng.Type) (zng.Type, error) {
+	if typ == cast || typ == zng.TypeNull {
+		return cast, nil
+	}
+	if isStringy(typ) {
+		if isStringy(cast) {
+			return cast, nil
+		}
+	} else if isInty(typ) {
+		if isInty(cast) {
+			return cast, nil
+		}
+	} else if isFloaty(typ) {
+		if isFloaty(cast) {
+			return cast, nil
+		}
+	}
+	return nil, fmt.Errorf("type mismatch: '%s' cannot be used as '%s'", typ.ZSON(), cast.ZSON())
+}
+
+func isStringy(typ zng.Type) bool {
```

Comment:
Can these is<Type> funcs be moved to zng/type.go alongside the Is<Type> funcs that are already present there?

Reply:
Good idea. Those existing functions already to what I need if I call ID() and add IdError to stringy.